File: sdk/typescript/src/blockchain/CurrencyChainClient.ts
// Currency Chain client for TypeScript SDK
// Handles payments, staking, rewards on currency chain

import axios, { AxiosInstance } from 'axios';

export enum CurrencyChainTxType {
  Payment = 'payment',
  Stake = 'stake',
  Unstake = 'unstake',
  Reward = 'reward',
  Slash = 'slash',
  Swap = 'swap',
}

export interface Wallet {
  user_id: string;
  balance: number;
  staked: number;
  rewards_pending: number;
}

export interface CurrencyChainTransaction {
  id: string;
  tx_type: CurrencyChainTxType;
  from: string;
  to?: string;
  amount: number;
  status: 'pending' | 'confirmed' | 'failed';
  confirmations: number;
  block_height: number;
  created_at: number;
}

export class CurrencyChainClient {
  private client: AxiosInstance;

  constructor(rpcUrl: string = 'http://localhost:8546') {
    this.client = axios.create({
      baseURL: rpcUrl,
      headers: { 'Content-Type': 'application/json' },
    });
  }

  async createWallet(userId: string, initialBalance: number): Promise<Wallet> {
    const response = await this.client.post('/currency/create_wallet', {
      user_id: userId,
      initial_balance: initialBalance,
    });
    return response.data.wallet;
  }

  async getWallet(userId: string): Promise<Wallet> {
    const response = await this.client.get(`/currency/wallet/${userId}`);
    return response.data;
  }

  async getBalance(userId: string): Promise<number> {
    const wallet = await this.getWallet(userId);
    return wallet.balance;
  }

  async transfer(from: string, to: string, amount: number): Promise<CurrencyChainTransaction> {
    const response = await this.client.post('/currency/transfer', {
      from,
      to,
      amount,
      timestamp: Math.floor(Date.now() / 1000),
    });
    return response.data;
  }

  async stake(userId: string, amount: number, lockDurationSeconds: number): Promise<CurrencyChainTransaction> {
    const response = await this.client.post('/currency/stake', {
      user_id: userId,
      amount,
      lock_duration_seconds: lockDurationSeconds,
      timestamp: Math.floor(Date.now() / 1000),
    });
    return response.data;
  }

  async claimRewards(userId: string): Promise<CurrencyChainTransaction> {
    const response = await this.client.post('/currency/claim_rewards', {
      user_id: userId,
      timestamp: Math.floor(Date.now() / 1000),
    });
    return response.data;
  }

  async getUserTransactions(userId: string): Promise<CurrencyChainTransaction[]> {
    const response = await this.client.get(`/currency/transactions/${userId}`);
    return response.data;
  }

  async getTransaction(txId: string): Promise<CurrencyChainTransaction | null> {
    try {
      const response = await this.client.get(`/currency/transaction/${txId}`);
      return response.data;
    } catch (error: any) {
      if (error.response?.status === 404) {
        return null;
      }
      throw error;
    }
  }

  async waitForConfirmation(
    txId: string,
    confirmations: number = 6,
    maxWaitMs: number = 30000,
  ): Promise<CurrencyChainTransaction> {
    const startTime = Date.now();
    while (Date.now() - startTime < maxWaitMs) {
      const tx = await this.getTransaction(txId);
      if (tx && tx.confirmations >= confirmations) {
        return tx;
      }
      await new Promise(resolve => setTimeout(resolve, 1000));
    }
    throw new Error(`Transaction ${txId} did not confirm within ${maxWaitMs}ms`);
  }
}

export default CurrencyChainClient;

File: sdk/dart/lib/src/blockchain/currency_chain_client.dart
/// Currency Chain client for Dart SDK
/// Handles payments, staking, rewards on currency chain

import 'dart:convert';
import 'package:http/http.dart' as http;
import 'client.dart';

/// Currency Chain transaction types
enum CurrencyChainTxType {
  payment,
  stake,
  unstake,
  reward,
  slash,
  swap,
}

/// Wallet information
class Wallet {
  final String userId;
  final int balance;
  final int staked;
  final int rewardsPending;

  Wallet({
    required this.userId,
    required this.balance,
    required this.staked,
    required this.rewardsPending,
  });

  factory Wallet.fromJson(Map<String, dynamic> json) {
    return Wallet(
      userId: json['user_id'] as String,
      balance: json['balance'] as int,
      staked: json['staked'] as int,
      rewardsPending: json['rewards_pending'] as int,
    );
  }
}

/// Currency Chain client for economics operations
class CurrencyChainClient extends BlockchainClient {
  /// Create new currency chain client
  CurrencyChainClient({
    String rpcUrl = 'http://localhost:8546',
    String? wsUrl,
  }) : super(
    config: BlockchainConfig(
      rpcUrl: rpcUrl,
      wsUrl: wsUrl,
    ),
  );

  /// Create wallet for user
  Future<Wallet> createWallet(String userId, int initialBalance) async {
    final response = await http.post(
      Uri.parse('${config.rpcUrl}/currency/create_wallet'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'user_id': userId,
        'initial_balance': initialBalance,
      }),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to create wallet: ${response.body}');
    }

    final data = jsonDecode(response.body) as Map<String, dynamic>;
    return Wallet.fromJson(data['wallet']);
  }

  /// Get wallet balance
  Future<Wallet> getWallet(String userId) async {
    final response = await http.get(
      Uri.parse('${config.rpcUrl}/currency/wallet/$userId'),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to get wallet: ${response.body}');
    }

    final data = jsonDecode(response.body) as Map<String, dynamic>;
    return Wallet.fromJson(data);
  }

  /// Transfer tokens between users
  Future<Map<String, dynamic>> transfer(
    String from,
    String to,
    int amount,
  ) async {
    final response = await http.post(
      Uri.parse('${config.rpcUrl}/currency/transfer'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'from': from,
        'to': to,
        'amount': amount,
        'timestamp': DateTime.now().millisecondsSinceEpoch ~/ 1000,
      }),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to transfer tokens: ${response.body}');
    }

    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  /// Stake tokens for rewards
  Future<Map<String, dynamic>> stake(
    String userId,
    int amount,
    int lockDurationSeconds,
  ) async {
    final response = await http.post(
      Uri.parse('${config.rpcUrl}/currency/stake'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'user_id': userId,
        'amount': amount,
        'lock_duration_seconds': lockDurationSeconds,
        'timestamp': DateTime.now().millisecondsSinceEpoch ~/ 1000,
      }),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to stake tokens: ${response.body}');
    }

    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  /// Claim rewards
  Future<Map<String, dynamic>> claimRewards(String userId) async {
    final response = await http.post(
      Uri.parse('${config.rpcUrl}/currency/claim_rewards'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'user_id': userId,
        'timestamp': DateTime.now().millisecondsSinceEpoch ~/ 1000,
      }),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to claim rewards: ${response.body}');
    }

    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  /// Get user transaction history
  Future<List<Map<String, dynamic>>> getUserTransactions(String userId) async {
    final response = await http.get(
      Uri.parse('${config.rpcUrl}/currency/transactions/$userId'),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to get transactions: ${response.body}');
    }

    final data = jsonDecode(response.body) as List<dynamic>;
    return data.cast<Map<String, dynamic>>();
  }
}
